examples/quantum_variation.py

```diff
@@ -0,0 +1,327 @@
+from scipy.linalg import expm
+import numpy as np
+import inspect
+import tensorcircuit as tc
+import random
+import math
+import matplotlib.pyplot as plt
+
+tc.set_backend("tensorflow")
+
+# calculate the matirx of kth qubit exert matrix[[a, b], [c, d]]
+def up_to_matrixx(k, a, b, c, d):
+    I2 = np.array([[1, 0], [0, 1]]) * (1 + 0j)
+    K = np.array([[a, b], [c, d]]) * (1 + 0j)
+    um = I2
+    if k == 0:
+        um = K
+    for i in range(1, N):
+        if i == k:
+            um = np.kron(um, K)
+        else:
+            um = np.kron(um, I2)
+    return um
+
+
+# realize R gates in paper
+def R_gate(k, c):
+    if door[k][0] == 0:
+        c.rx(door[k][1] + 1, theta=ODE_theta[k])
+    if door[k][0] == 1:
+        c.ry(door[k][1] + 1, theta=ODE_theta[k])
+    if door[k][0] == 2:
+        c.rz(door[k][1] + 1, theta=ODE_theta[k])
+    if door[k][0] == 3:
+        c.rxx(door[k][1] + 1, door[k][2] + 1, theta=ODE_theta[k])
+    if door[k][0] == 4:
+        c.ryy(door[k][1] + 1, door[k][2] + 1, theta=ODE_theta[k])
+    if door[k][0] == 5:
+        c.rzz(door[k][1] + 1, door[k][2] + 1, theta=ODE_theta[k])
+    if door[k][0] == 6:
+        c.crx(door[k][1] + 1, door[k][2] + 1, theta=ODE_theta[k])
+    if door[k][0] == 7:
+        c.cry(door[k][1] + 1, door[k][2] + 1, theta=ODE_theta[k])
+    if door[k][0] == 8:
+        c.crz(door[k][1] + 1, door[k][2] + 1, theta=ODE_theta[k])
+
+
+# realize U gates in paper
+def U_gate(k, c):
+    if door[k][0] == 0:
+        c.cx(0, door[k][1] + 1)
+    if door[k][0] == 1:
+        c.cy(0, door[k][1] + 1)
+    if door[k][0] == 2:
+        c.cz(0, door[k][1] + 1)
+    if door[k][0] == 3:
+        c.multicontrol(
+            0, door[k][1] + 1, door[k][2] + 1, ctrl=[0], unitary=tc.gates._xx_matrix
+        )
+    if door[k][0] == 4:
+        c.multicontrol(
+            0, door[k][1] + 1, door[k][2] + 1, ctrl=[0], unitary=tc.gates._yy_matrix
+        )
+    if door[k][0] == 5:
+        c.multicontrol(
+            0, door[k][1] + 1, door[k][2] + 1, ctrl=[0], unitary=tc.gates._zz_matrix
+        )
+    if door[k][0] == 6:
+        c.multicontrol(
+            0,
+            door[k][1] + 1,
+            door[k][2] + 1,
+            ctrl=[0, door[k][1] + 1],
+            unitary=tc.gates._x_matrix,
+        )
+    if door[k][0] == 7:
+        c.multicontrol(
+            0,
+            door[k][1] + 1,
+            door[k][2] + 1,
+            ctrl=[0, door[k][1] + 1],
+            unitary=tc.gates._y_matrix,
+        )
+    if door[k][0] == 8:
+        c.multicontrol(
+            0,
+            door[k][1] + 1,
+            door[k][2] + 1,
+            ctrl=[0, door[k][1] + 1],
+            unitary=tc.gates._z_matrix,
+        )
+
+
+# realize Hamilton gates in ancillary circuit
+def H_gate(q, c):
```

Comment:
the def of H gate seems identical as the method for U gate?

Reply:
then why keep two, what is the difference?

Reply:
only to diff the parameter of simulation circuit and hamilton